In [ ]:
import os
# Configure session cookie for Advent of Code
# os.environ['AOC_SESSION']

## DAY 1

### Part 1

You arrive at the secret entrance to the North Pole base ready to start decorating. Unfortunately, the password seems to have been changed, so you can't get in. A document taped to the wall helpfully explains:

"Due to new security protocols, the password is locked in the safe below. Please see the attached document for the new combination."

The safe has a dial with only an arrow on it; around the dial are the numbers 0 through 99 in order. As you turn the dial, it makes a small click noise as it reaches each number.

The attached document (your puzzle input) contains a sequence of rotations, one per line, which tell you how to open the safe. A rotation starts with an L or R which indicates whether the rotation should be to the left (toward lower numbers) or to the right (toward higher numbers). Then, the rotation has a distance value which indicates how many clicks the dial should be rotated in that direction.

So, if the dial were pointing at 11, a rotation of R8 would cause the dial to point at 19. After that, a rotation of L19 would cause it to point at 0.

Because the dial is a circle, turning the dial left from 0 one click makes it point at 99. Similarly, turning the dial right from 99 one click makes it point at 0.

So, if the dial were pointing at 5, a rotation of L10 would cause it to point at 95. After that, a rotation of R5 could cause it to point at 0.

The dial starts by pointing at 50.

You could follow the instructions, but your recent required official North Pole secret entrance security training seminar taught you that the safe is actually a decoy. The actual password is the number of times the dial is left pointing at 0 after any rotation in the sequence.

For example, suppose the attached document contained the following rotations:

L68
L30
R48
L5
R60
L55
L1
L99
R14
L82

Following these rotations would cause the dial to move as follows:

    The dial starts by pointing at 50.
    The dial is rotated L68 to point at 82.
    The dial is rotated L30 to point at 52.
    The dial is rotated R48 to point at 0.
    The dial is rotated L5 to point at 95.
    The dial is rotated R60 to point at 55.
    The dial is rotated L55 to point at 0.
    The dial is rotated L1 to point at 99.
    The dial is rotated L99 to point at 0.
    The dial is rotated R14 to point at 14.
    The dial is rotated L82 to point at 32.

Because the dial points at 0 a total of three times during this process, the password in this example is 3.

Analyze the rotations in your attached document. What's the actual password to open the door?

In [ ]:

import aocd

In [ ]:
puzzle = aocd.get_puzzle(year=2025,day=1)

In [ ]:
puzzle.input_data

'R1\nL31\nR5\nL11\nL37\nL46\nR31\nR36\nL33\nR37\nL35\nR15\nR15\nL9\nR40\nL45\nR5\nR35\nR40\nL12\nR46\nR5\nR10\nL47\nR49\nL47\nL50\nR11\nR26\nL7\nL24\nR21\nL9\nR27\nR28\nL45\nR15\nL13\nR39\nL32\nL11\nL32\nR29\nL48\nL41\nR14\nR26\nL5\nL35\nR72\nL52\nL71\nR99\nL28\nL71\nR33\nR69\nR90\nR8\nR26\nR78\nL93\nL28\nL83\nR18\nR27\nL73\nL72\nL14\nR26\nR85\nL81\nL48\nR89\nL57\nR60\nL60\nL17\nL34\nR51\nR81\nL81\nR87\nR13\nL47\nL6\nL47\nL18\nL82\nL15\nR8\nR47\nL40\nL61\nL70\nL49\nL36\nR45\nL57\nR34\nR94\nR577\nR23\nR25\nL725\nL47\nL53\nR37\nR63\nL308\nL292\nL54\nL46\nL19\nL81\nR680\nR20\nL23\nR23\nR513\nR12\nR77\nR8\nR390\nR18\nR82\nR68\nL68\nL366\nL34\nR109\nL98\nL75\nR64\nR972\nR628\nR781\nR442\nR77\nR64\nL95\nL39\nR870\nL266\nR12\nL40\nL96\nR90\nR16\nR9\nR75\nR19\nL19\nL91\nL31\nR47\nL45\nL96\nL84\nL43\nL18\nL20\nL19\nL10\nL90\nR82\nR66\nR70\nR139\nL57\nR120\nL61\nR41\nR78\nR23\nR944\nR5\nL77\nR27\nR94\nR76\nL970\nR47\nR42\nR74\nL831\nR928\nR612\nR28\nL81\nR53\nR28\nL76\nR93\nL117\nL78\nL22\nL1\nL

In [ ]:
xs = puzzle.examples[0].input_data.split()
xs

['L68', 'L30', 'R48', 'L5', 'R60', 'L55', 'L1', 'L99', 'R14', 'L82']

In [ ]:
x = xs[0]
x[0], int(x[1:])

('L', 68)

In [ ]:

t =  int(x[1:])

In [ ]:

5 % 3

2

In [ ]:

l = [1,2,3]
l[5 % 3]

3

In [ ]:
(10 - 5) % 100

5

In [ ]:
# L10 instr
(5 - 10 ) % 100

95

In [ ]:
# dial value - delta % 100
(50 - t) % 100

82

In [ ]:
x[0], int(x[1:])

('L', 68)

In [ ]:

t =  int(x[1:])
t

68

In [ ]:
# dial value - delta % 100 if L otherwise dial value + delta % 100
(82 - t) % 100

14

In [ ]:
v = 50 
c = 0
for x in xs:
    signal, delta = x[0], (int(x[1:]))
    if signal=='L': v = (v - delta) % 100
    else:  v = (v + delta) % 100
    print(v)
    if v == 0: c+=1
assert c == 3


82
52
0
95
55
0
99
0
14
32


In [ ]:
def part1(xs):
    v,c = 50,0
    for x in xs:
        signal,delta = x[0],int(x[1:])
        v = (v-delta)%100 if signal=='L' else (v+delta)%100
        if v==0: c+=1
    return c

part1(xs)

3

In [ ]:
xs = puzzle.input_data.split()
part1(xs)


1021

### Part 2 

You're sure that's the right password, but the door won't open. You knock, but nobody answers. You build a snowman while you think.

As you're rolling the snowballs for your snowman, you find another security document that must have fallen into the snow:

"Due to newer security protocols, please use password method 0x434C49434B until further notice."

You remember from the training seminar that "method 0x434C49434B" means you're actually supposed to count the number of times any click causes the dial to point at 0, regardless of whether it happens during a rotation or at the end of one.

Following the same rotations as in the above example, the dial points at zero a few extra times during its rotations:

    The dial starts by pointing at 50.
    The dial is rotated L68 to point at 82; during this rotation, it points at 0 once.
    The dial is rotated L30 to point at 52.
    The dial is rotated R48 to point at 0.
    The dial is rotated L5 to point at 95.
    The dial is rotated R60 to point at 55; during this rotation, it points at 0 once.
    The dial is rotated L55 to point at 0.
    The dial is rotated L1 to point at 99.
    The dial is rotated L99 to point at 0.
    The dial is rotated R14 to point at 14.
    The dial is rotated L82 to point at 32; during this rotation, it points at 0 once.

In this example, the dial points at 0 three times at the end of a rotation, plus three more times during a rotation. So, in this example, the new password would be 6.

Be careful: if the dial were pointing at 50, a single rotation like R1000 would cause the dial to point at 0 ten times before returning back to 50!

Using password method 0x434C49434B, what is the password to open the door?

Okay since on though this example does this mean that I have to count every time there is a pass through zero regardless if it's exactly stopping at zero? 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Exactly! In Part 2, you're counting *every time* the dial passes through 0 during the rotation, not just when it stops at 0.

So if you're at position 95 and do R60, the dial goes: 95 → 96 → 97 → 98 → 99 → **0** → 1 → 2 → ... → 54 → 55. That's one pass through 0 during the rotation, plus you end at 55 (not 0), so you count 1 total for that move.

The tricky part is figuring out: given a starting position, a direction (L or R), and a distance, how many times does the dial click past 0?

Do you want to think about how you might calculate that, or would you like a hint about the approach?

Let's work through logic slowly from the example starting at 0 and going left 68 . My idea is given the starting number that the dial is currently at in the direction either left or right, the ending number. If it's bigger than the starting number, it means past zero. If you're in the right direction, if it's smaller than the starting number, it also means past zero. After applying the delta 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good thinking! Let's test your logic with the first example:

- Start: 50
- Instruction: L68
- End: (50 - 68) % 100 = 82

So we start at 50 and end at 82. Since we're going **left** (decreasing), and 82 > 50, that means we wrapped around through 0.

Let me check if I understand your rule correctly:
- Going **right** (R): if end < start, we passed through 0
- Going **left** (L): if end > start, we passed through 0

Is that what you're thinking? Let's test it on another example to see if it holds up. What about the third instruction in the example: starting at 52, doing R48, ending at 0?

In [ ]:
v = 50
xs = puzzle.examples[0].input_data.split()
x = xs[0]
signal,delta = x[0],int(x[1:])
v1 = (v-delta)%100 if signal=='L' else (v+delta)%100
v1,v


(82, 50)

Something on the terms in this case: v1 will be bigger than v. As the signal was left that means increment the count for 0s by one 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Yes! You've got it for that case. When going left (L) and v1 > v, we crossed through 0 at least once.

Now let's check your rule with the R48 example:
- Start: 52
- Instruction: R48  
- End: (52 + 48) % 100 = 0

Here, we're going **right** and v1 (0) < v (52). According to your rule, that would mean we passed through 0. But actually, we land exactly at 0 without wrapping around (52 → 53 → ... → 99 → 0). 

So does this case count as "passing through" or just "landing on"? What do you think?

In [ ]:
v = 52
x = "R48"
signal,delta = x[0],int(x[1:])
v1 = (v-delta)%100 if signal=='L' else (v+delta)%100
v1,v


(0, 52)

This will still work because in this case v1 is smaller than v and therefore which is counted as one. Exactly landing would not do anything because it still counts as a decrease 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Perfect! You're right. Whether we land exactly on 0 or pass through it, the logic still works:

- Going **right** (R): if v1 < v, count it (we either passed through or landed on 0)
- Going **left** (L): if v1 > v, count it (we either passed through or landed on 0)

But there's one more thing to consider: what if we pass through 0 **multiple times** in a single rotation? For example, if you're at position 10 and go L150, you'd wrap around more than once.

Do you think that's possible given the constraints of this problem, or will each rotation cross 0 at most once?

In [ ]:
xs = puzzle.input_data.split()
xs[:100]
# lets find the maximum number in the puzzle
max_num = max(int(x[1:]) for x in xs)
max_num


999

It's definitely possible. I've just looked at the input data 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good catch! So with deltas up to 999, you could definitely wrap around multiple times. For example, if you're at 10 and go L250, you'd pass through 0 twice.

So how would you calculate the number of times you cross 0 for a given rotation? 

Here's a hint: think about how many "full laps" around the dial (0-99) you're making. Does that give you an idea?

In [ ]:
v = 52
x = "L251"
signal,delta = x[0],int(x[1:])
v1 = (v-delta)%100 if signal=='L' else (v+delta)%100
v1,v


(1, 52)

In [ ]:
5 // 2

2

In [ ]:
delta // 100

2

So is this I think this is the way I have for the calc of a full laps. Let's do the integer division of delta per 100 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great start! `delta // 100` gives you the number of complete laps around the dial.

But there's a bit more to consider: even if you don't make a complete lap, you might still cross 0 once depending on where you start and where you end up.

For example:
- Start at 52, go L251: You make 2 complete laps (that's 2 crossings), plus you go from 52 to 1, which crosses 0 once more. Total: 3 crossings.
- Start at 10, go L5: You make 0 complete laps, but you still cross from 10 to 5, which crosses 0 once.

So you need: `(delta // 100)` **plus** something to account for the partial lap. 

Can you figure out what condition would add 1 more for that partial crossing?

In [ ]:
delta % 100

51

You made a mistake there. As you started at 52, you go left at 251. You made two complete laps with two crossings. But then they go from 52 to 1 which does not cross zero once more. So total two crossings 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

You're absolutely right! I made an error. Let me reconsider:

- Start at 52, go L251
- 2 complete laps = 2 crossings through 0
- The remaining 51 steps left from 52 takes you to 1, which doesn't cross 0
- Total: 2 crossings

So the full laps give you `delta // 100` crossings, and then you need to check if the partial movement (the remainder) causes an additional crossing based on your starting position.

For going **left** from position v with a remainder of `delta % 100`, when would that cross 0?

We need to take the delta per 100 and add that as increments for laps, Regardless of the signal if you're doing an integer division for 100 and if that number comes up bigger than zero, you've got to add that to the counter. Then apply modulo 100 To the delta itself. If it's below 100, it's not going to make a difference; it's going to return the same value. If it's above, you just get the remainder which is what we want to apply the left and right logic as mentioned before. If v1 is bigger or smaller than 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Excellent! You've got the complete logic:

1. Add `delta // 100` to the counter (for complete laps, regardless of direction)
2. Apply `delta % 100` as the actual movement
3. Then use your original logic: 
   - If going **left** (L) and v1 > v: add 1
   - If going **right** (R) and v1 < v: add 1

So the total count for each rotation would be:
```
count = (delta // 100) + (1 if condition_based_on_direction else 0)
```

Want to try implementing this in your `part2` function?

In [ ]:
xs = puzzle.examples[0].input_data.split()


position = 50
zero_count = 0
for move in xs:
    direction, dist = move[0], int(move[1:])
    lap_crossings = dist // 100
    zero_count += lap_crossings # distances > 99
    step = dist % 100 # brings back to 99 and below steps
    previous = position
    if direction == 'L':
        destination = (position - step) % 100
        if destination > previous: zero_count += 1 # went down pass zero or equal to 0
        position = destination
    elif direction == 'R':
        destination = (position + step) % 100
        if destination < previous: zero_count += 1 # went over or equal to 0
        position = destination

zero_count

6

In [ ]:
def part2(xs):
    position = 50
    zero_count = 0
    for move in xs:
        direction, dist = move[0], int(move[1:])
        lap_crossings = dist // 100 # over 99 add laps
        zero_count += lap_crossings
        step = dist % 100 # this brings up to the <= 99 L or R
        previous = position # save position
        if direction == 'L':
            destination = (position - step) % 100
            if destination > previous: zero_count += 1 # this mean i lapped included 0 landed
            position = destination
        elif direction == 'R':
            destination = (position + step) % 100
            if destination < previous: zero_count += 1 # as above
            position = destination
    return zero_count

In [ ]:
xs = puzzle.examples[0].input_data.split()
xs

xs = puzzle.input_data.split()
part1(xs), part2(xs)


(1021, 5956)

OK my answer is too high so there are edge cases here that I am not accounting for Let's investigate and discuss step-by-step what happens when step equals to zero  so  destination = previous, lets see if this woudl trigger a count

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good thinking! Let's trace through what happens when `step == 0`:

If `step = 0` (meaning `delta % 100 == 0`), then:
- `destination = previous` (no movement on the partial step)
- For **left** (L): `destination > previous` would be False (since they're equal)
- For **right** (R): `destination < previous` would be False (since they're equal)

So when `step == 0`, you wouldn't add 1 for the partial movement, which seems correct.


Lets work this case as it might be an issue as i am overcounting. But i think just for the left. Lets go through a L and R example step by step.  

I think it might be linked to a start at zero positon. 

- case 1 i go left 10. "i think i might be counting this and should not as this was counted n previous loop (landed on zero)"
- case 2 i go right 10. "i will not count this"

Please be critical and verify my claims 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠

Let's trace through carefully:

**Case 1: Start at 0, go L10**
- previous = 0
- destination = (0 - 10) % 100 = 90
- Check: destination (90) > previous (0)? **YES** → count += 1

But did we actually cross 0? Going left from 0: 0 → 99 → 98 → ... → 90. We *started* at 0 but didn't *cross through* 0. So this is an **overcount**!

**Case 2: Start at 0, go R10**
- previous = 0
- destination = (0 + 10) % 100 = 10
- Check: destination (10) < previous (0)? **NO** → no count

This is correct.

---

Your intuition is spot on! When starting at position 0 and going **left**, your logic incorrectly adds 1. The fix would be: if `previous == 0`, don't add 1 for the partial crossing.

Does that match what you were thinking?

In [ ]:
def part2(xs):
    position = 50
    zero_count = 0
    for move in xs:
        direction, dist = move[0], int(move[1:])
        lap_crossings = dist // 100
        zero_count += lap_crossings
        step = dist % 100
        previous = position
        if direction == 'L':
            destination = (position - step) % 100
            if (destination > previous or destination == 0) and previous != 0: zero_count += 1 # as before but dont count if previous is zero and going LEFT!
            position = destination
        elif direction == 'R':
            destination = (position + step) % 100
            if destination < previous and previous != 0: zero_count += 1
            position = destination
    return zero_count

In [ ]:

xs = puzzle.examples[0].input_data.split()
part2(xs)

6

In [ ]:
xs = puzzle.input_data.split()
part1(xs), part2(xs)


(1021, 5933)

Okay this has worked. So I think this was the bug. If I was going left and I was in zero as a previous position, the logic would count when that should not be the case. It had been counted before when landed on zero (previous loop). This was not obvious. 
Funny enough this was not the case when the position was right and this wasn't immediately obvious to me at all. I had to break down the edge cases one by one. For example bigger than 99 as a delta, step equals to zero, which doesn't seem to be a problem in this particular case because you'll land on the destination equal to the position . And finally if the position is 0 and we're going left and rught, how that affects the counter logic as I designed it 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Excellent debugging work! You systematically identified the edge cases and found the subtle bug. Let me summarize what you discovered:

**The key insight:** When `previous == 0` and going **left**, your wrap-around logic (`destination > previous`) incorrectly counted a crossing, even though you started exactly at 0 and didn't actually pass through it during the move.

**Why it didn't happen going right:** When at 0 going right, `destination` will always be ≥ 0, so `destination < previous` is never true (since previous = 0). No false positive!

This is a great example of how modular arithmetic edge cases can be tricky. The asymmetry between left and right movements around 0 wasn't immediately obvious, and you had to test specific scenarios to catch it.

Nice work breaking it down methodically! Does this solution feel solid to you now, or are there other edge cases you want to verify?

In [ ]:

-11 % 100, -11%20, 10%100

(89, 9, 10)

Okay we use modulo operator here. Is this behavior of negative numbers in the modulo operator a mathematical definition? I'm struggling to see why it works as a difference from left to right 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Great question! The behavior of the modulo operator with negative numbers is actually *language-specific*, not purely mathematical.

In Python, `(-11) % 100 = 89` because Python's modulo always returns a non-negative result with the same sign as the divisor (100 in this case). It's defined so that: `a == (a // b) * b + (a % b)`.

So for `-11 % 100`:
- `-11 // 100 = -1` (floor division)
- `-11 % 100 = 89` (because `-1 * 100 + 89 = -11`)

This is why your "going left" logic works! When you do `(position - step) % 100`, if the subtraction goes negative, Python automatically wraps it around to the correct positive value (0-99 range).

**Why the asymmetry?** Going left can produce negative intermediate values (which wrap around), while going right produces positive values that might exceed 99 (also wrapping). The wrap-around at 0 behaves differently in each direction because of how the subtraction vs addition interacts with the modulo.

Does that help clarify why the left/right cases needed different handling?

#So giving you more right examples left and right. How is this would always wrap around 0 to 99 range? 

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good question! Let's trace through some concrete examples to see how Python's modulo keeps everything in the 0-99 range:

**Going Right (addition):**
- Start at 95, go R10: `(95 + 10) % 100 = 105 % 100 = 5` ✓
- Start at 0, go R150: `(0 + 150) % 100 = 50` ✓

**Going Left (subtraction):**
- Start at 5, go L10: `(5 - 10) % 100 = (-5) % 100 = 95` ✓
- Start at 50, go L68: `(50 - 68) % 100 = (-18) % 100 = 82` ✓

The key is that Python's `%` operator handles both:
- Positive overflow (when addition goes > 99): wraps back from 0
- Negative values (when subtraction goes < 0): wraps back from 99

Both always give you a result in the range [0, 99].

Would you like to test a few more examples yourself to see the pattern?